storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -16,44 +16,2351 @@
+                                if (appendInfo.firstOrLastOffsetOfFirstBatch() < localLog.logEndOffset()) {
+                                    // we may still be able to recover if the log is empty
+                                    // one example: fetching from log start offset on the leader which is not batch aligned,
+                                    // which may happen as a result of AdminClient#deleteRecords()
+                                    boolean hasFirstOffset = appendInfo.firstOffset() != UnifiedLog.UNKNOWN_OFFSET;
+                                    long firstOffset = hasFirstOffset ? appendInfo.firstOffset() : records.batches().iterator().next().baseOffset();
+
+                                    String firstOrLast = hasFirstOffset ? "First offset" : "Last offset of the first batch";
+                                    List<String> offsets = new ArrayList<>();
+                                    for (Record record : records.records()) {
+                                        offsets.add(String.valueOf(record.offset()));
+                                        if (offsets.size() == 10) break;
+                                    }
+                                    throw new UnexpectedAppendOffsetException(
+                                            "Unexpected offset in append to " + topicPartition() + ". " + firstOrLast + " " +
+                                                    appendInfo.firstOrLastOffsetOfFirstBatch() + " is less than the next offset " + localLog.logEndOffset() + ". " +
+                                                    "First 10 offsets in append: " + String.join(", ", offsets) + ", last offset in" +
+                                                    " append: " + appendInfo.lastOffset() + ". Log start offset = " + logStartOffset,
+                                            firstOffset, appendInfo.lastOffset());
+                                }
+                            }
+
+                            // update the epoch cache with the epoch stamped onto the message by the leader
+                            validRecords.batches().forEach(batch -> {
+                                if (batch.magic() >= RecordBatch.MAGIC_VALUE_V2) {
+                                    assignEpochStartOffset(batch.partitionLeaderEpoch(), batch.baseOffset());
+                                } else {
+                                    // In partial upgrade scenarios, we may get a temporary regression to the message format. In
+                                    // order to ensure the safety of leader election, we clear the epoch cache so that we revert
+                                    // to truncation by high watermark after the next leader election.
+                                    if (leaderEpochCache.nonEmpty()) {
+                                        LOG.warn("Clearing leader epoch cache after unexpected append with message format v{}", batch.magic());
```

Comment:
logger

Reply:
I'm switching all the `LOG` in not static methods to `logger`.